### DISCUSSION TITLE
Collection Display - Cards

### DISCUSSION THREAD

**Discussion by Kematia** [directus/directus#5679] [Feature Requests]
Hello bandits!

I've come across a couple use cases where the Collection Display Template, while being great for inline content, is not suited for _richer_ content types. 

## The Problem
The first example that comes to mind is a "Product", which has a _cover image_, a _title_, _subtitle_ and a _rich text area_, with some additional information like a _price_  or _category_.

In the current setup, it works alright if you have it setup with a cover image and a price, but gets messy real quick when you start to add the _price_ or _category_ fields. You can remove fields, but it will be harder for editors to spot duplicates, missing content, and generally keep track of where they are.

![image](https://user-images.githubusercontent.com/17851386/118274037-f14ee000-b4c4-11eb-8fdc-01268f2507d2.png)

## The Solution (?)

![image](https://user-images.githubusercontent.com/17851386/118277181-cb2b3f00-b4c8-11eb-96d1-92d7ab004d34.png)
_An ad interim solution for one of my clients_

I propose we add another type of collection template display, called `Display Card`.
The card would expose some preset _slots_ which you can "link" to a collection item property.

I was thinking of 4 main _slots_: `title`, `subtitle`, `excerpt`, `image`

## Options

The Display Card would have the following options: 

### Layout

Controls if the card is horizontal or vertical. An horizontal card could be used in the `list-*` interfaces, while a vertical card could be used when hovering over an item (sort of a preview).

### Show Actions
A simple boolean to control whether to show the CRUDD actions. If you're in a `list-*` interface you'd want to show them, while you'd avoid them on a `v-table`.

## The Setup

![image](https://user-images.githubusercontent.com/17851386/118276126-92d73100-b4c7-11eb-8e90-3d4711e6e578.png)
_Behold, the results of 5 years of art school..._

The interface to setup the card would be a simple grid which mocks the layout of the card, where the user can click and select which fields to "link", similar to the current Display Template setup.

For example, if the user indicates no `title`, the card could show up like this:

![image](https://user-images.githubusercontent.com/17851386/118276312-cc0fa100-b4c7-11eb-87d0-c267ece5ec12.png)

Or, if missing both the `image`, `title` and `subtitle`:

![image](https://user-images.githubusercontent.com/17851386/118276403-e8abd900-b4c7-11eb-8d1b-9bba3a75147a.png)

## 80/20

I think that, while being a _fairly opinionated_ graphical representation, it could be worthwhile to implement. It could be a good companion to the _Display Template_, one core use could be to preview a `directus_user`.

**Comment by benhaynes** [comment]
I like this, and thank you for the research, @Kematia !! I do like the idea of a more image dedicated Display/Interface... but this raises an interesting point! While you're trying to adjust/improve the "display", this is actually the "interface" (m2m and/or o2m). But this makes me wonder if there is a way to have more complicated displays that inject into an interface. That would allow us to keep the simplicity of a single core M2M, but inject more complicated displays such as this. @rijkvanzanten ... anything promising in this way if thinking? Not sure if it will work since it would mean one display for multiple fields. Might be unnecessary complexity.

**Reply by rijkvanzanten** [reply]
That would be one way of doing it! Another would be to have 2 display templates per relational interface row (bit of a concession in between, but gets us 75% of the way there), a third approach could be to allow for multi-line display templates 🤔 

In the design example above, it kinda looks like the row itself holding the ID is what we have right now, where the box with the "rich preview" is the example below. @benhaynes we were already discussing this GitHub style (/ Wikipedia style) preview box on hover, maybe whatever config we use for that could be used for this rich preview within the relational "card" rows style? And the relational interfaces could have a "accordion: start open" option to always render the rich preview. (We'll have to keep size in mind, if you have 25 of those pretty rows, you'll have to scroll for like 3km before you get to the next field)

**Reply by benhaynes** [reply]
I think we should expand the collection "Display" to allow prioritizing the _type_ of data (Data, Long Form Text, Image, Location, etc) a bit more. Being able to configure a hover preview card seems like a great Core feature, and then that same view could be used on things like the M2M on hover.

We could also include this as logic in the Interface itself to show the accordion slightly differently. We don't need to cover all UIs, that's why we have Interface extensions, but since we combined `Images` into the M2M, it'd be good to have this option.

I really like the two Wikipedia examples, and can imagine using that _image_ area for a small map, prominent date, etc. I'm wondering if this can use the existing Display Template, or if we should add a separate configuration for this multiline/card style.

I like what @Kematia proposed (`Behold, the results of 5 years of art school...` 😆 ), with all/most sections being optional. Some inputs (eg: `title`) could accept a template string... or that could even _become_ the Display Template. 🤔

**Reply by Kematia** [reply]
> I think we should expand the collection "Display" to allow prioritizing the _type_ of data (Data, Long Form Text, Image, Location, etc) a bit more. Being able to configure a hover preview card seems like a great Core feature, and then that same view could be used on things like the M2M on hover.

The key point I think is to have the dynamic templates existing on a collection basis, and the tools for the users to set those up in the backend. After that is just developing display interfaces to solve discrete _presentation_ problems, be it _Core_ or _Custom_. 

For example - once you have the template, _Core_ can have a card-like interface for list-* interfaces (rows), and another (maybe core) interface for the hover preview card, etcetera. Then I (_user_) can set up my _own Custom interface_ to display my custom preview with the same set of data/template, or spin up another _custom template_ with the backend interface.

In the same way we use a _collection-builder_ to setup collection fields, we could have a drag n' drop/WYSIWYG interface to set those templates up. 

One use case would be to describe a front-end preview card, which could be different from an _Admin App_ preview card.

> We could also include this as logic in the Interface itself to show the accordion slightly differently. We don't need to cover all UIs, that's why we have Interface extensions, but since we combined `Images` into the M2M, it'd be good to have this option.
> 
> I really like the two Wikipedia examples, and can imagine using that _image_ area for a small map, prominent date, etc. I'm wondering if this can use the existing Display Template, or if we should add a separate configuration for this multiline/card style.

Maybe we could think about an extendible system? Collection Displays? Thinking on my feet we could have:

- Inline (currently Display Template)
- Multiline (maybe redundant with _Card_)
- Card (Rich card with images, actions, etc)
- Row (for use in v-table)
- Link (wikipedia-like links, where you can assign another Collection Display to show on hover when using a rich text editor)
- Custom

This is just an example on how we could use a dynamic system to _present_  different and discrete collection "previews" in many places of the App.

> I like what @Kematia proposed (`Behold, the results of 5 years of art school...` 😆 ), with all/most sections being optional. Some inputs (eg: `title`) could accept a template string... or that could even _become_ the Display Template. 🤔

I personally really like the simplicity of the Display Template, and I don't know if it would be worth the effort to refactor it everywhere in the application - I do think that the `title` and `subtitle`, etc fields should work the same as the current Display Template (click and assign the related field from the dropdown).

**Reply by Kematia** [reply]
https://github.com/directus/directus/discussions/5639 for example could be resolved by a _custom display template_ as described above.

**Reply by benhaynes** [reply]
This all makes sense, although I do want to make sure this doesn't get overengineered. What I'm thinking is that we have a handful of inputs at the Collection-level that allow for configuring all of this. However, just as Directus decouples presentation from data, I'd like to do the same thing here. So in essence, there would be some set of fields, like:

* **Title** — Could _maybe_ use the existing Display Template, since those are both inline... though there might be a good reason to keep those separate.
* **Kicker / Subtitle** — Just need to decide if this is going above (kicker) or below (subtitle) the title.
* **Body** — Longer form text. Would be interesting to allow multiple fields here with _new-lines_ between, but that might not be needed.
* **Image / Location** — Optional image or location field. We could base the preview on the type, showing either a thumbnail or small map preview. Not sure what other media could be shown here.

Having the above, we could then generate the preview cards, enhance interface lists, etc. I think this is a good MVP.

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
